tests/integration/applications/continuous_write_charm/src/charm.py

```diff
@@ -0,0 +1,229 @@
+from ops.model import ActiveStatus, Relation, WaitingStatus
+from pymongo import MongoClient
+from tenacity import RetryError, Retrying, stop_after_delay, wait_fixed
+
+logger = logging.getLogger(__name__)
+
+DATABASE_NAME = "continuous_writes_database"
+COLLECTION_NAME = "continuous_writes_collection"
+REPLICATION_COLL_NAME = "test_ubuntu_collection"
+PEER = "application-peers"
+PROC_PID_KEY = "proc-pid"
+LAST_WRITTEN_FILE = "last_written_value"
+
+
+class ContinuousWritesApplication(CharmBase):
+    """Application charm that continuously writes to MongoDB."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        # Charm events
+        self.framework.observe(self.on.start, self._on_start)
+
+        self.framework.observe(
+            self.on.clear_continuous_writes_action, self._on_clear_continuous_writes_action
+        )
+        self.framework.observe(
+            self.on.start_continuous_writes_action, self._on_start_continuous_writes_action
+        )
+        self.framework.observe(
+            self.on.stop_continuous_writes_action, self._on_stop_continuous_writes_action
+        )
+
+        # Database related events
+        self.database = DatabaseRequires(self, "database", DATABASE_NAME)
+        self.framework.observe(self.database.on.database_created, self._on_database_created)
+
+    # ==============
+    # Properties
+    # ==============
+
+    @property
+    def _peers(self) -> Relation | None:
+        """Retrieve the peer relation (`ops.model.Relation`)."""
+        return self.model.get_relation(PEER)
+
+    @property
+    def app_peer_data(self) -> dict:
+        """Application peer relation data object."""
+        if self._peers is None:
+            return {}
+
+        return self._peers.data[self.app]
+
+    @property
+    def _database_config(self) -> dict[str, str]:
+        """Returns the database config to use to connect to the MongoDB cluster."""
+        # In some tests we want to write directly to mongos, but the config-server does not
+        # support integrations to client applications, so the data to connect is set via config.
+        if not (data := list(self.database.fetch_relation_data().values())):
+            return {"uris": self.model.config.get("mongos-uri", None)}
+
+        data = data[0]
+        username, password, endpoints, replset, uris = (
+            data.get("username"),
+            data.get("password"),
+            data.get("endpoints"),
+            data.get("replset"),
+            data.get("uris"),
+        )
+
+        if None in [username, password, endpoints, replset, uris]:
+            return {}
+
+        return {
+            "user": username,
+            "password": password,
+            "endpoints": endpoints,
+            "replset": replset,
+            "uris": uris,
+        }
+
+    # ==============
+    # Helpers
+    # ==============
+
+    def _start_continuous_writes(self, starting_number: int, db_name: str, coll_name: str) -> None:
```

Comment:
For all subsequent methods
```suggestion
    def _start_continuous_writes(self, starting_number: int, db_name: str, collection_name: str) -> None:
```